### ISSUE TITLE
Use custom classes in fireducks

### ISSUE THREAD

**Issue by weidinger-c** [fireducks-dev/fireducks#13]
Hi, I have a problem using fireducks:
I have a pandas dataframe which also contains a custom class as column type. 
This custom class is just a bunch of strings and some member functions (which is why i not only added the plain strings to the dataframe)
When using Fireducks i get the error:

RuntimeError: ArrowInvalid: ('Could not convert 45112_54290 45112_54290.tile 45112_54290.metadata with type TileInfo: did not recognize Python value type when inferring an Arrow data type', 'Conversion failed for column __fireducks_dummy_0 with type object')

At:
  pyarrow/error.pxi(91): pyarrow.lib.check_status

my custom class looks like this:

```
class TileInfo:
    def __init__(
        self, tile_name: Path, tile_filepath: Path, tile_metadata_filepath: Path
    ):
        self.tile_name = tile_name
        self.tile_filepath = tile_filepath
        self.tile_metadata_filepath = tile_metadata_filepath

    def _read_metadata(self) -> dict:
        return _read_tile_metadata(self.tile_metadata_filepath)

    def _write_metadata(self, metadata: dict) -> bool:
        return _write_tile_metadata(self.tile_metadata_filepath, metadata)

    def __str__(self):
        return f"{self.tile_name} {self.tile_filepath} {self.tile_metadata_filepath}"

    def __repr__(self):
        return f"{self.tile_name} {self.tile_filepath} {self.tile_metadata_filepath}"

    tile_name = str
    tile_filepath = Path
    tile_metadata_filepath = Path
```

**Comment by qsourav** [comment]
@weidinger-c, thank you for your query.

DataFrame in FireDucks currently does not support a column of custom classes due to performance. Since the backend implementation is backed by smarter data structures and implemented in C++, as long as your columnar data can be accessed as a consecutive memory, we can promise better performance. We are currently working on supporting custom class cases as well (by falling back to pandas, instead of throwing an error), but you may not expect speedup from such cases.

By the way, I can see that you are probably using a custom class of three string objects. Will it be possible to treat it as three string-typed columns instead and do the processing? You should be able to perform tasks using string accessor methods (e.g., s.str.upper() etc.) and since those are vectorized implementation, both pandas and FireDucks will perform better I believe.

**Comment by weidinger-c** [comment]
Hi @qsourav, I have now removed the custom class and added it as string directly. But now I have got another error:

NotImplementedError: Can not convert arrow::Scalar to fireducks::Scalar: uint16

It seems that there is no Pandas Dataframe feature parity and unsigned intergers are not supported?

**Comment by qsourav** [comment]
@weidinger-c, thank you so much for your kind update.

The unsigned types are actually supported for common operations like sort, filter etc. 
You are probably performing some aggregate operation (sum, max, etc.) on a uint16 column, and you are facing this issue due to an internal bug. I am marking it as a bug and will fix it in our next release. We appreciate your kind reporting!

 - [] support of columns having custom class  objects
 - [] bug in performing aggregation on unsigned columns (Can not convert arrow::Scalar to fireducks::Scalar: uint16)

**Comment by weidinger-c** [comment]
@qsourav yes it happens for min() and max() operations

**Comment by qsourav** [comment]
By the way, would you like to tell us the methods you intend to perform on the column of custom class objects? Although supporting them would not be performance efficient, as mentioned earlier, we are working on this, so just to take note of the methods you are using.

**Comment by weidinger-c** [comment]
I now have removed the custom class. I am writing a lib that does spatial tiling of 3D pointclouds, and each tile has its filepath and metadata file. Each of these tiles with its metadata is in a pandas df, which contains name, min/max of all values of this tile etc.
I would not have processed the custom class any special, as it just contained some strings of filepaths and provided some loading and saving functions. 

What I am hoping with Fireducks it the speed up of min/max, filtering operations etc of the huge pointcloud tiles. Looking forward to the next release.

**Comment by qsourav** [comment]
Perfect! For the filtering, you might be able to do it without any problem now as well (please confirm if you face any other issues). 
For the min/max, just to check the performance, would you like to confirm by casting the uint16 to int32 (e.g., s.astype("int32"))?
I especially like to know if fixing this bug can help you to achieve your goal. If you face any other issues, please let me know.

**Comment by weidinger-c** [comment]
@qsourav I just created a bunch of test array and firefly so far was always faster.

Unfortunately I cannot test it in my code so far, because I stumbled on another error.

`RuntimeError: ArrowTypeError: ("Expected bytes, got a 'bool' object", 'Conversion failed for column __fireducks_dummy_0 with type object')`

**Comment by qsourav** [comment]
It seems like you are trying to perform some operation that involves somehow mixing a bool with a non-bool (string, maybe) type. Our ongoing support of custom class objects would probably address this situation as well, but would it be possible to share some more information so that we can debug your case? 

  - since you said "contains_active" is a boolean column, probably "row" is a dataframe. is it right?
  - what is the type of "tile_intersection"? a ndarray, list, or a Series? does it contain boolean data as well?
  - what is "idx", some numeric indices like 0, 4, 2, ..._ or some boolean mask like True, True, False etc.?

Also, since FireDucks follows a lazy-execution model, the actual error might be in some different location. 
If you set FIREDUCKS_FLAGS="--benchmark-mode," it will be an eager execution mode, and the actual location of this error might be spotted.

**Comment by weidinger-c** [comment]
Where can I set this flag FIREDUCKS_FLAGS="--benchmark-mode"?

**Comment by qsourav** [comment]
> Where can I set this flag FIREDUCKS_FLAGS="--benchmark-mode"?

It is an environment variable. You may like to set it before running your python program/tests, or you may like to set it within your program as follows (in that case, make sure it is done before importing FireDucks):
```
import os
os.environ["FIREDUCKS_FLAGS"] = "--benchmark-mode"
:
: 
import fireducks.pandas as pd
# your-program
```

**Comment by weidinger-c** [comment]
Hmm, with the flag it now happens at another place:

`for idx, row in self.tiles.iterrows():`

ArrowTypeError: ("Expected bytes, got a 'int' object", 'Conversion failed for column __fireducks_dummy_0 with type object')

**Comment by weidinger-c** [comment]
Would it help you if I export the pandas dataframe?

**Comment by weidinger-c** [comment]
Yes thats correct, here is a glimps whats inside the dataframe (a mix of strings, intergers, uint, floats, bools)

      tile_name  num_points  (x_tile, min)  (x_tile, max)  (y_tile, min)  (y_tile, max)  (z_tile, min)  (z_tile, max)  (point_id, min)  ...  (point_source_id, max)  (red, min)  (red, max)  (green, min)  (green, max)  (blue, min)  (blue, max)  (label, min)  (label, max)
0   45112_54290       16850      4511200.0      4511300.0      5429000.0      5429100.0           -inf            inf                0  ...                       0           0       65278             0         65278            0        65278            65            65
1   45113_54290       26130      4511300.0      4511400.0      5429000.0      5429100.0           -inf            inf            16850  ...                       0           0       65278             0         65278            0        65278            65            65

**Comment by weidinger-c** [comment]
e.g. i take only tiles in the dataframe that contain relevant points and then load the points and do filter stuff etc.
Could you provide an example how to iterate over them without iterrow?

**Comment by qsourav** [comment]
Do you want to perform something similar to follows, where tile_intersections is a numpy array of boolean values?
```
for idx, row in self.tiles.iterrows():
    row["contains_active"] = tile_intersections[idx]

filtered_tiles = self.tiles[self.tiles["contains_active"]]
```

**Comment by weidinger-c** [comment]
Typically I iterate over the rows (with represent the tile metadata) and then do some processign with functions

e.g.
```
for idx, row in self.tiles.iterrows():
            # Only disable active flags per point if the active state of the tile has changed (otherwise they are already set)
            if np.logical_xor(row["contains_active"], tile_intersections[idx]):
                self.active_flags[idx] = np.zeros(row["num_points"], dtype=bool)
            row["contains_active"] = tile_intersections[idx]
```

**Comment by weidinger-c** [comment]
len(self.tiles) == len(tile_intersections)
type(self.tiles.index): <class 'pandas.core.indexes.range.RangeIndex'>

**Comment by qsourav** [comment]
I am assuming 
  - "self.active_flags" is an array-of-array-like structure. 
  - When the xor result is False, it will be an empty array of boolean.

In that case, the data frame-related operations, like calculation of column-wise xor, assignment, etc., can be done without a loop, and we can limit the numpy array creation part within the loop. Assuming a sample input as follows:
```
tiles = pd.DataFrame()
tiles["contains_active"] = [True, False, False, True]
tiles["num_points"] = [10, 6, 8, 13]
tile_intersections = np.array([True, True, False, False])
active_flags = [None] * 4
```

maybe we can rewrite the loop as follows?
```
log_xor_op = tiles["contains_active"] ^ tile_intersections
N = tiles["num_points"].where(log_xor_op).fillna(0).astype(int)  # use num_point when xor result is "True", else 0
for idx, n in enumerate(N): 
  active_flags[idx] = np.zeros(n, dtype=bool)
tiles["contains_active"] = tile_intersections
```

I am not sure if the above will fully help you, but this way, we can surely avoid the overhead of iterating on each row of the input data frame (self.tiles), which is a very slow operation.
